chore: code documentation and usage (#23)
ory/elements: README.md

- Dynamically adapts the user interface to your Ory identity schema, sign-in and
  flow configuration

Elements supports integrating with:

- React
- Preact
- ExpressJs (experimental)

## Getting Started

Install Ory Elements into your existing React / Preact or ExpressJS application.
You can find example apps [here](#example-apps)

**React**

```shell
npm install @ory/elements --save
```

**Preact**

```shell
npm install @ory/elements-preact --save
```

**ExpressJs**

```shell
npm install @ory/elements-markup --save
```

## Example Apps

Dive into Ory Elements with our `react` and `preact` SPA example applications,
located in the `examples/` directory.

To run the example application you will need a couple things:

> 1. An [Ory Cloud (free) account](https://console.ory.sh/)

Author: yeah i'll fix that quick

Now you can see Ory Elements in action by opening http://localhost:3000 in your
browser!

## Quickstart: Storybook

Explore the Ory Elements via [Storybook](https://storybook.js.org/)!

Clone this repository and run:

```shell
npm run initialize
npm run build
# or `npm run build:clean` to ensure no packages have cached versions
npm run storybook
```

## Contributing

Write a new component inside the `src/react-components` directory with its
corresponding css in `src/theme`. Check it out by writing a new story for the
component in the `src/stories` folder.

Add a test to verify the component works correctly by creating a new file next
to the component file with the same name and an added `*.spec.ts` extension. All
tests are written in [Playwright](https://playwright.dev/) for component tests
and E2E tests.

**Example Apps**

To contribute an example application, please add it to the `examples/` folder.
To ensure the example works correctly within the Lerna build system, add the
`elements` package to the example `package.json` with an asterisk `*` as the
version.

Below is an example of how you should add the pacakge.

```json
...

> "@ory/elements": "*",

Author: alright, yeah makes sense. This section is mostly for contributions

> "@ory/elements": "*",

Reviewer: Ah, okay, I see...

...
```

## Understanding Elements

### Bundling System

Elements uses [Lerna](https://lerna.js.org/) to bundle each package in the
Elements mono-repository. This also helps with package management and build

> caching. Lerna also publishes the code to the npm for us.

Reviewer: ```suggestion
caching. Lerna also publishes the code to the public [npm registry](https://www.npmjs.com/) for us.
```

      size,
      color,
    }),
  card: UserAuthCard({
    title: !(flow.refresh || flow.requested_aal === "aal2")
      ? "Sign In"
      : "Two-Factor Authentication",
    flow: flow as SelfServiceFlow,
    flowType: "login",
    cardImage: "ory-logo.svg",
    additionalProps: {
      forgotPasswordURL: "recovery",
      signupURL: initRegistrationUrl,
      logoutURL: logoutUrl,
    },
  }),
})
```

---

### Component System

> Ory Elements solely rely on React components since they are easy to write and

Reviewer: Heh, this is a weird one. 

- "Ory Elements solely relies on..." (as a whole product)
- "Ory Elements solely rely on..." (as their individual parts)

I stumbled upon this when reading, because I am so used to elements being our complete suite of components. But in the bigger picture they really are more of a suite of individual elements, so the second version makes sense. Weird one! :)

Reviewer: Actually below we are saying "Ory Elements is a component library" which implies they're a library instead of individual parts. So we should be consistent here and use the first case, IMO:

```suggestion
Ory Elements solely relies on React components since they are easy to write and
```

provides support to a large React based ecosystem. The project then bundles
these components to their respective needs. An example is bundling for Preact
which you can find in the
[packages](https://github.com/ory/elements/tree/main/packages) folder. It uses

> the react components directly in this case but bundles it specifically for

Reviewer: ```suggestion
the React components directly in this case but bundles it specifically for
```

      size,
      color,
    }),
  card: UserAuthCard({
    title: !(flow.refresh || flow.requested_aal === "aal2")
      ? "Sign In"
      : "Two-Factor Authentication",
    flow: flow as SelfServiceFlow,
    flowType: "login",
    cardImage: "ory-logo.svg",
    additionalProps: {
      forgotPasswordURL: "recovery",
      signupURL: initRegistrationUrl,
      logoutURL: logoutUrl,
    },
  }),
})
```

---

### Component System

Elements solely rely on React components since they are easy to write and
provides support to a large React based ecosystem. The project then bundles
these components to their respective needs. An example is bundling for Preact
which you can find in the
[packages](https://github.com/ory/elements/tree/main/packages) folder. It uses
the react components directly in this case but bundles it specifically for
Preact support.

> Each component relies on a some CSS, which are located in the

Reviewer: ```suggestion
Each component relies on some CSS styles, which are located in the
```

[theme](https://github.com/ory/elements/tree/main/src/theme) directory. To
understand how this works, please refer to the [CSS System](#css-system)

> #### ExpressJs systems

Reviewer: ```suggestion
#### Express JS systems
```

Reviewer: The correct spelling seems to be Express JS, apparently.

> Express is an edge-case which requires the component to be wrapped by
> `ReactDOMServer` to produce only HTML. Each component needs to be wrapped by

Reviewer: ```suggestion
Express JS has limited support for React components, by only rendering the markup to HTML without the dynamic elements of a component. Each component needs to be wrapped by
```

Author: hmm, i mean it is still "dynamic" - it just isn't dynamic on the client side. So the props can be injected still on server-side and the html produced will still change.

Reviewer: Hmm, yes, you're right... What's a good word for all the hooks, then?

`ComponentWrapper` which essentially uses `ReactDOMServer`. The
`elements-markup` package then bundles the React library with it so that the
React code lives with the component library.

Here is an example of exporting the `UserAuthCard`.

```ts
import {
  UserAuthCard as userAuthCard,
  UserAuthCardProps,
} from "../react-components"

export const UserAuthCard = (props: UserAuthCardProps) => {
  return ComponentWrapper(userAuthCard(props))
}

export type { UserAuthCardProps } from "../react-components"
```

### Asset System

> Assets are bundled into a singular `style.css` file under each pacakges `dist/`

Reviewer: ```suggestion
Assets are bundled into a singular `style.css` file under each packages `dist/`
```

I do the same typo, way too often :D It's a weird word to type.

folder. Anything placed inside the

> [assets](https://github.com/ory/elements/tree/main/src/assets) folder will be

Reviewer: Do we have tree shaking enabled in these parts?

Not really related to this PR, just out of curiosity. It sounds like we are just copying all the files in this folder to one CSS files.

Author: they get bundled through Vite with minification. But yeah we need to do some sort of tree-shaking or chunking later. Since you most likely don't want all fonts and all styles at the same time.

bundled. They can also be directly imported by the React components to be used
and are sometimes required by a component. An example is the
[Social Button Component](https://github.com/ory/elements/blob/main/src/react-components/button-social.tsx#L11-L13).

## Versioning and Publishing

> Elements uses a mixture of manual and automated publishing. We first need to

Author: I guess Ory Elements is better then. But I didn't want to sound repetitive :)
